src/typechecker.rs

```diff
@@ -298,6 +349,79 @@ impl<'a> Typechecker<'a> {
+                variable,
+                range,
+                block,
+            } => {
+                self.typecheck_node(range);
+
+                // We don't need to typecheck variable after this
+                let var_id = self
+                    .compiler
+                    .var_resolution
+                    .get(&variable)
+                    .expect("missing resolved variable");
+                if let Type::List(type_id) = self.type_of(range) {
+                    self.variable_types[var_id.0] = type_id;
+                    self.set_node_type_id(variable, type_id);
+                } else {
+                    self.variable_types[var_id.0] = ANY_TYPE;
+                    self.set_node_type_id(variable, ERROR_TYPE);
+                    self.error("For loop range is not a list", range);
+                }
+
+                self.typecheck_node(block);
+                if self.type_id_of(block) != NONE_TYPE {
+                    self.error("Blocks in looping constructs cannot return values", block);
+                }
+
+                self.set_node_type_id(node_id, NOTHING_TYPE);
+            }
+            AstNode::While { condition, block } => {
+                self.typecheck_node(block);
+                if self.type_id_of(block) != NONE_TYPE {
+                    self.error("Blocks in looping constructs cannot return values", block);
+                }
+
+                self.typecheck_node(condition);
+
+                // the condition should always evaluate to a boolean
+                if self.type_of(condition) != Type::Bool {
+                    self.error("The condition for while loop is not a boolean", condition);
+                    self.set_node_type_id(node_id, ERROR_TYPE);
+                } else {
+                    self.set_node_type_id(node_id, self.type_id_of(block));
+                }
+            }
+            AstNode::Match {
+                ref target,
+                ref match_arms,
+            } => {
+                // Check all the output types of match
+                let output_types = self.typecheck_match(target, match_arms);
+                match output_types.len().cmp(&1) {
+                    Ordering::Greater => {
+                        self.oneof_types.push(output_types);
+                        self.set_node_type(
+                            node_id,
+                            Type::OneOf(OneOfId(self.oneof_types.len() - 1)),
+                        );
+                    }
+                    Ordering::Equal => {
+                        self.set_node_type_id(
+                            node_id,
+                            *output_types
+                                .iter()
+                                .next()
+                                .expect("Will contain one element"),
+                        );
+                    }
+                    Ordering::Less => {
+                        self.set_node_type_id(node_id, NONE_TYPE);
```

Comment:
Since `match` is an expression, I think this should be `nothing` because you can assign expressions to values. `none` belongs to statements.